linode-launch.py

```diff
@@ -5,184 +5,107 @@
-            swap = filter(lambda d: d[u'LABEL'] == u'swap', current_disks)
-            if swap:
-                swap = swap[0]
-                assert(swap[u'SIZE'] == int(swap_size))
-                assert(swap[u'TYPE'] == u'swap')
-                disks.append({u'DiskID': swap[u'DISKID']})
-            else:
-                disks.append(client.linode_disk_create(LinodeID = node[u'LinodeID'], Label = u'swap', Type = u'swap', Size = swap_size))
-            raw = filter(lambda d: d[u'LABEL'] == u'raw', current_disks)
-            if raw:
-                raw = raw[0]
-                assert(raw[u'SIZE'] == int(raw_size))
-                assert(raw[u'TYPE'] == u'raw')
-                disks.append({u'DiskID': raw[u'DISKID']})
-            elif raw_size > 0:
-                disks.append(client.linode_disk_create(LinodeID = node[u'LinodeID'], Label = u'raw', Type = u'raw', Size = raw_size))
-            disks = [unicode(d[u'DiskID']) for d in disks]
-            disklist = u','.join(disks)
-            logging.info("%s", "{} disks: {}".format(label, disks))
-            ceph_config = filter(lambda c: c[u'Label'] == u'ceph', configs)
-            if ceph_config:
-                logging.info("%s", "{} ceph config already setup".format(label))
-                config = ceph_config[0]
-            else:
-                config = client.linode_config_create(LinodeID = node[u'LinodeID'], Label = u'ceph', KernelID = kernel, Disklist = disklist, RootDeviceNum = 1)
-        except linapi.ApiError as err:
-            raise
-
-        if active and active[0][u'STATUS'] == 1:
-            logging.info("%s", "Linode {} already running.".format(label))
-        else:
-            client.linode_boot(LinodeID = node[u'LinodeID'], ConfigID = config[u'ConfigID'])
-            logging.info("%s", "booted {}: {}".format(label, node[u'LinodeID']))
-        linodes.append({"id": node[u'LinodeID'], "name": name, "label": label, "ip_private": ip_private, "ip_public": ip_public, "group": machine['group'], "user": "root", "key": SSH_PRIVATE_KEY_FILE})
-
-def create(*args, **kwargs):
-    try:
-        do_create(*args, **kwargs)
-    except Exception as e:
-        logging.exception(e)
-        os._exit(1)
-
-def launch(client):
-    datacenters = client.avail_datacenters()
-    plans = client.avail_linodeplans()
-    distributions = client.avail_distributions()
-    kernels = client.avail_kernels()
-
-    datacenter = filter(lambda d: d[u'LOCATION'].lower().find(DATACENTER.lower()) >= 0, datacenters)[0][u'DATACENTERID']
-    distribution = filter(lambda d: d[u'LABEL'].lower().find(DISTRIBUTION.lower()) >= 0, distributions)[0][u'DISTRIBUTIONID']
-    kernel = filter(lambda k: k[u'LABEL'].lower().find(str(KERNEL).lower()) >= 0, kernels)[0][u'KERNELID']
-
-    running = client.linode_list()
-
-    with closing(ThreadPool(50)) as pool:
```

Comment:
In my experience it isn't necessary at least with linode_api4, because the create call and the delete call just start creation and deletion, they don't block until it's finished.  We have to wait anyway (linode-wait.py, linode-destroy.py).  It's a lot simpler to maintain non-multi-threaded code.  I'll wait and see when I get to 100 linodes (just got the limit raised on my account) how well this works though, you could be right.